webhook/resources/backuptarget/validator.go

```diff
@@ -0,0 +1,304 @@
+	admission.DefaultValidator
+	ds *datastore.DataStore
+}
+
+func NewValidator(ds *datastore.DataStore) admission.Validator {
+	return &backupTargetValidator{ds: ds}
+}
+
+func (b *backupTargetValidator) Resource() admission.Resource {
+	return admission.Resource{
+		Name:       "backuptargets",
+		Scope:      admissionregv1.NamespacedScope,
+		APIGroup:   longhorn.SchemeGroupVersion.Group,
+		APIVersion: longhorn.SchemeGroupVersion.Version,
+		ObjectType: &longhorn.BackupTarget{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Create,
+			admissionregv1.Update,
+			admissionregv1.Delete,
+		},
+	}
+}
+
+func (b *backupTargetValidator) Create(request *admission.Request, newObj runtime.Object) error {
+	backupTarget := newObj.(*longhorn.BackupTarget)
+
+	if !util.ValidateName(backupTarget.Name) {
+		return werror.NewInvalidError(fmt.Sprintf("invalid name %v", backupTarget.Name), "")
+	}
+
+	if err := b.validateBackupTargetURL(backupTarget.Spec.BackupTargetURL); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	if err := b.validateCredentialSecret(backupTarget.Spec.CredentialSecret); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	if err := b.handleAWSIAMRoleAnnotation(backupTarget, nil); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	return nil
+}
+
+func (b *backupTargetValidator) Update(request *admission.Request, oldObj runtime.Object, newObj runtime.Object) error {
+	oldBackupTarget := oldObj.(*longhorn.BackupTarget)
+	newBackupTarget := newObj.(*longhorn.BackupTarget)
+
+	urlChanged := oldBackupTarget.Spec.BackupTargetURL != newBackupTarget.Spec.BackupTargetURL
+	secretChanged := oldBackupTarget.Spec.CredentialSecret != newBackupTarget.Spec.CredentialSecret
+
+	if urlChanged {
+		if err := b.validateBackupTargetURL(newBackupTarget.Spec.BackupTargetURL); err != nil {
+			return werror.NewInvalidError(err.Error(), "")
+		}
+	}
+
+	if secretChanged {
+		if err := b.validateCredentialSecret(newBackupTarget.Spec.CredentialSecret); err != nil {
+			return werror.NewInvalidError(err.Error(), "")
+		}
+	}
+
+	if urlChanged || secretChanged {
+		if err := b.validateDRVolume(newBackupTarget); err != nil {
+			return werror.NewInvalidError(err.Error(), "")
+		}
+
+		if err := b.handleAWSIAMRoleAnnotation(newBackupTarget, oldBackupTarget); err != nil {
+			return werror.NewInvalidError(err.Error(), "")
+		}
+	}
+
+	if oldBackupTarget.Spec.Default && !newBackupTarget.Spec.Default {
+		backupTargetMap, err := b.ds.ListBackupTargetsRO()
+		if err != nil {
+			return werror.NewInvalidError(err.Error(), "")
+		}
+		isDefaultBackupTargetChanged := false
+		for _, bt := range backupTargetMap {
+			if bt.Name != newBackupTarget.Name && bt.Spec.Default {
+				isDefaultBackupTargetChanged = true
+				break
+			}
+		}
+		if !isDefaultBackupTargetChanged {
+			return werror.NewInvalidError("prohibit disabling a default backup target directly", "")
+		}
+	}
```

Comment:
Is it possible that there is another BackupTarget has spec.Default=true and the current BackupTarget is changing from true to false? Doesn't that means there are two BackupTarget Default=true?

Can we just **disable user to update spec.Default from true to false**, 
**this should only be achieved by the controller when one BackupTarget become default and others got forced to change from true to false**

So we can simply
```
if oldBackupTarget.Spec.Default && !newBackupTarget.Spec.Default {
     return werror.NewInvalidError("prohibit disabling a default backup target directly", "")
}
```

Reply:
>Is it possible that there is another BackupTarget has spec.Default=true and the current BackupTarget is changing from true to false? Doesn't that means there are two BackupTarget Default=true?

Yes, when users set another backup target default, two backup targets have the default value true in the Spec but there is only one backup target that has `Status.Default` true and that's why we would get the `Status.Default`, not `Spec.Default` to check if it is a default backup target.

It might be blocked by simply checking `Spec.Default` changed because the backup target controller updates the spec of the backup target and it still would run through the validator.

Yeah, it should be as simple as your comment, and we should allow the controller to update backup target CRs.
I will modify the implementation by this comment.

Reply:
Modified and added an annotation to check if the updated comes from the controller.  Thanks.